### DISCUSSION TITLE
Typescript - `decodeJwt` & `jwtVerify` (and maybe others) should take in a generic for strongly typed payloads

### DISCUSSION THREAD

**Discussion by MrSimmmons** [panva/jose#568] [Ideas]
Pretty much the title. I have a scenario where I have custom data in the payload of the JWT, and I have to force the type of the output of both of these methods.

My issue, is that I have to disable my linter for that line because the linter doesn't like forcing types (which is generally best practice) and I'm not a _huge_ fan of having to do that because it kind of defeats the purpose of a linter.
```ts
// eslint-disable-next-line ...
const payload = jose.decodeJwt(token) as jose.JWTPayload & { foo: string };
```

So, I'm proposing to add a generic to these methods so linters can be happy

```ts
const payload = jose.decodeJwt<{ foo: string }>(token);
```

**Comment by panva** [comment]
Rather than relying on generics and blindly trusting what you receive I believe it's far better to have a validation function for the payload that a) validates the payload is actually of your expected type and in doing so b) uses type predicate to allow you continue with your expected type.

```ts
interface Foo extends jose.JWTPayload {
  foo: string
}

function isFoo(payload: jose.JWTPayload): payload is Foo {
  return 'foo' in payload && typeof payload.foo === 'string'
}

const { payload } = await jose.jwtVerify(token, key)

if (isFoo(payload)) {
  payload.foo // :tada:
}
```

**Reply by MrSimmmons** [reply]
While yes that would work, IMO its not very scaleable as the more custom key / value pairs you add just make the predicate check arbitrarily long. And _especially_ complicated if one of the things that you are passing in the jwt is a nested object.

What about wrapping the provide generic in `Partial`. That way you can still "force" the user to check for `undefined` cases if they are required to do so

**Reply by panva** [reply]
Of course you can also omit the validation in the predicate if you want to blindly trust the shape of the received payload.

**Reply by aryzing** [reply]
@MrSimmmons you can easily validate with a schema library such as Zod.

**Comment by aryzing** [comment]
@panva how come `decodeJwt` doesn't use generics but `jwtVerify` does?

```ts
export declare function jwtVerify<PayloadType = JWTPayload>(jwt: string | Uint8Array, key: KeyLike | Uint8Array, options?: JWTVerifyOptions): Promise<JWTVerifyResult<PayloadType>>;

export declare function decodeJwt(jwt: string): JWTPayload;
```

**Reply by panva** [reply]
How come? I just forgot to add it there.

**Reply by panva** [reply]
https://github.com/panva/jose/releases/tag/v5.1.0

**Comment by ruhil6789** [comment]
Hi — I've built @pq-jose which implements this for Node.js today. @panva @aryzing @ItalyPaleAle @
Inviting all above to review and share your feedback please

npm install @pq-jose

It covers:
- SignJWT / jwtVerify (ML-DSA-65, FIPS 204) — 3-part signed token
- EncryptJWT / jwtDecrypt (ML-KEM-768, FIPS 203) — 5-part JWE token  
- exportJWK / importJWK (AKP key type, draft-ietf-cose-dilithium)
- createLocalJWKSet / createRemoteJWKSet

API is intentionally similar to this library. 55 tests passing.

https://www.npmjs.com/package/@pq-jose

https://github.com/pq-jwt/PQ-JOSE

https://pq-jwt.github.io/

Happy to answer questions or collaborate on interoperability.
